colexec: wrap unsupported post process specs and ON expressions

asubiotto opened the pull request:
Unsupported post process specs and ON expressions are now planned after a
columnar operator by wrapping a noop processor that evaluates the unsupported
expression before returning the row.

This commit also makes sure that two types are Equivalent before planning any
comparison or binary expression. Returning an error if not.

Release note (bug fix): mixed type comparison or binary expressions that could
have previously returned wrong results in the vectorized execution engine fall
back to using the row execution engine.

Fixes #45626

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45724)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45724)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colexec/execplan.go, line 195 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1bIPbN99mqt6f7MoBj:-M1bIPbN99mqt6f7MoBk:bnag3io) ([raw file](https://github.com/cockroachdb/cockroach/blob/406cd8f64f0ea31fa9683e223e441c321534c71c/pkg/sql/colexec/execplan.go#L195)):*
> ```Go
> 	case core.HashJoiner != nil:
> 		if !core.HashJoiner.OnExpr.Empty() && core.HashJoiner.Type != sqlbase.InnerJoin {
> 			return false, errors.Newf("can't plan vectorized outer hash joins with ON expressions")
> ```

nit: I think technically, LeftSemi and LeftAnti are not "outer" joins, so I'd keep "non-inner" wording.
___
*[pkg/sql/colexec/execplan.go, line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1bIobzAdhf3z_4oGpw:-M1bIobzAdhf3z_4oGpx:b-yaykbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/406cd8f64f0ea31fa9683e223e441c321534c71c/pkg/sql/colexec/execplan.go#L391)):*
> ```Go
> 
> // createAndWrapRowSource takes a processor spec, creating the row source and
> // wrapping it using wrapRowSource. Note that the post process spec is included
> ```

nit: `s/wrapRowSource/wrapRowSources/g`.
___
*[pkg/sql/colexec/execplan.go, line 436 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1bJs7RAdJqUrURFCRG:-M1bJs7RAdJqUrURFCRH:b-c3utyd) ([raw file](https://github.com/cockroachdb/cockroach/blob/406cd8f64f0ea31fa9683e223e441c321534c71c/pkg/sql/colexec/execplan.go#L436)):*
> ```Go
> 				)
> 			}
> 			// The wrapped processors need to be passed the post-process specs,
> ```

This comment referred to setting `post` variable of `NewColOperaror` to empty, so I think it should be moved outside of this function in there as well.
___
*[pkg/sql/colexec/execplan.go, line 821 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1bKMknBBYPH32_DbC9:-M1bKMknBBYPH32_DbCA:b-7qxiwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/406cd8f64f0ea31fa9683e223e441c321534c71c/pkg/sql/colexec/execplan.go#L821)):*
> ```Go
> 				if core.MergeJoiner.Type != sqlbase.JoinType_INNER {
> 					return result, errors.AssertionFailedf(
> 						"ON expression (%s) was unexpectedly planned for merge joiner with join type",
> ```

Missing on `%s` argument in the format string.
___
*[pkg/sql/colexec/execplan.go, line 863 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1bKSbdEKwp-KuGyGtr:-M1bKSbdEKwp-KuGyGts:b1d8tyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/406cd8f64f0ea31fa9683e223e441c321534c71c/pkg/sql/colexec/execplan.go#L863)):*
> ```Go
> 				// TODO(yuzefovich): remove this when we're confident in ON expression
> 				//  support.
> 				result.IsStreaming = false
> ```

I think we want to remove this now that we don't have any special code used for ON expressions.
___
*[pkg/sql/colexec/execplan.go, line 1077 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1bKxOd8bDTa_bec4lL:-M1bKxOd8bDTa_bec4lM:b316e95) ([raw file](https://github.com/cockroachdb/cockroach/blob/406cd8f64f0ea31fa9683e223e441c321534c71c/pkg/sql/colexec/execplan.go#L1077)):*
> ```Go
> }
> 
> // planPostProcessSpec plans the post processing stage specified in post. On top
> ```

nit: `s/post. On/post on/g`.
___
*[pkg/sql/logictest/testdata/logic_test/vectorize, line 1145 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1bLF4p0Cn0JgQUMWlf:-M1bLF4p0Cn0JgQUMWlg:b-rb031g) ([raw file](https://github.com/cockroachdb/cockroach/blob/406cd8f64f0ea31fa9683e223e441c321534c71c/pkg/sql/logictest/testdata/logic_test/vectorize#L1145)):*
> ```
> # Test that unsupported post process specs get wrapped in the vectorized engine.
> statement ok
> CREATE TABLE mixed_type_a (a INT, b TIMESTAMPTZ);
> ```

nit: semicolons are not necessary.
___
*[pkg/sql/logictest/testdata/logic_test/vectorize, line 1166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1bLKx9D_8jwDZLtwds:-M1bLKx9D_8jwDZLtwdt:b-owogjv) ([raw file](https://github.com/cockroachdb/cockroach/blob/406cd8f64f0ea31fa9683e223e441c321534c71c/pkg/sql/logictest/testdata/logic_test/vectorize#L1166)):*
> ```
> false
> 
> # Merge join on expressions also get wrapped.
> ```

nit: I would `s/on/ON/g`.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
Updated to wrap the whole tablereader if post is unsupported since we kind of need to given that the materializer doesn't know how to handle unneeded columns. This is also better for performance.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-:-M1g15nEB7eYvI0HQ7pS:bihamf3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colexec/execplan.go, line 436 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1bJs7RAdJqUrURFCRG:-M1eqwc11GYmj2wmjvEQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/406cd8f64f0ea31fa9683e223e441c321534c71c/pkg/sql/colexec/execplan.go#L436)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This comment referred to setting `post` variable of `NewColOperaror` to empty, so I think it should be moved outside of this function in there as well.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/execplan.go, line 821 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1bKMknBBYPH32_DbC9:-M1erSNsEVm-Og2Lzj1-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/406cd8f64f0ea31fa9683e223e441c321534c71c/pkg/sql/colexec/execplan.go#L821)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Missing on `%s` argument in the format string.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/execplan.go, line 863 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1bKSbdEKwp-KuGyGtr:-M1erXj71-Daqak4tDKK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/406cd8f64f0ea31fa9683e223e441c321534c71c/pkg/sql/colexec/execplan.go#L863)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think we want to remove this now that we don't have any special code used for ON expressions.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto commented:
This PR also wraps cases like `SELECT a in (1, 2) FROM t` where `a` is an int and is being compared to a tuple. I think this might be overly restrictive but it might be good to be cautious.

The check is currently restrictive in that it denies any mixed-type comparison. I'm not sure I want to add special casing to it though. The other option is to make it a blacklist and disallow if we're doing mixed type comparisons that also include a set of types (these would be timestamp, timestamp tz, date, and probably interval?)

Thoughts?

yuzefovich reviewed (commented):
I would probably add a special whitelist case for `SELECT ... IN ...` because we have an optimized operator for it. I'm worried that if we introduce the blacklist, it could be a source of bugs, but if somehow screw up whitelisting of "select in", then in the worst case, we would fallback to rowexec and the performance would suffer (but no correctness issues as could be in the opposite direction).

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-:-M1gStcY-wft3u6eVswO:b23yjak)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colexec/execplan.go, line 521 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1gReJV7lw09I-7d7LX:-M1gReJW-D3HeZfHzPCx:bet8y8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de332a25bbbe9343e41e7a54a069b3d55fa2623/pkg/sql/colexec/execplan.go#L521)):*
> ```Go
> 	result.IsStreaming = false
> 
> 	// resultCopyPreSpecPlanningStateShallowCopy is a shallow copy of the result
> ```

nit: `s/resultCopyPre/resultPre/g`.
___
*[pkg/sql/colexec/execplan.go, line 1091 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1gSSPN9DzGCxC8bI3a:-M1gSSPN9DzGCxC8bI3b:bhzd9gu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de332a25bbbe9343e41e7a54a069b3d55fa2623/pkg/sql/colexec/execplan.go#L1091)):*
> ```Go
> 				inputTypes = append(inputTypes, input.ColumnTypes)
> 			}
> 			result.resetToState(ctx, resultPreSpecPlanningStateShallowCopy)
> ```

I don't think we're creating any monitors or accounts when planning `TableReader`, right? We probably don't need to reset the result. Where is this coming from?


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
Done, added whitelist for other specialized operators too although I disallow them for dates, timestamp(tz), and interval. The code I added seems to not disallow anything that we already support that is specified in `testoptlogic`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-:-M1ghnnT9igdAy3rN2DH:b9pka8i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colexec/execplan.go, line 1091 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1gSSPN9DzGCxC8bI3a:-M1g_4cc-XOEl5P1Phpw:bigwkqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de332a25bbbe9343e41e7a54a069b3d55fa2623/pkg/sql/colexec/execplan.go#L1091)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I don't think we're creating any monitors or accounts when planning `TableReader`, right? We probably don't need to reset the result. Where is this coming from?
</blockquote></details>

We do because of the operator and result column types. The monitors and accounts are added for completeness. We could technically just do `result = resultPre...` but that's more brittle if anybody decides to add more state to the result that is not reset properly.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 11 of 11 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45724)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colexec/execplan.go, line 1091 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1gSSPN9DzGCxC8bI3a:-M1kQgx15shhbdBgzvUM:bdo3vm3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de332a25bbbe9343e41e7a54a069b3d55fa2623/pkg/sql/colexec/execplan.go#L1091)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

We do because of the operator and result column types. The monitors and accounts are added for completeness. We could technically just do `result = resultPre...` but that's more brittle if anybody decides to add more state to the result that is not reset properly.
</blockquote></details>

I see, so the actual reason is that we need to reset `.Op` and `.ColumnTypes` field to original values. And I see the value of having `reset` function right next to the struct. SGTM.
___
*[pkg/sql/colexec/execplan.go, line 505 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1kSKdBCpzn5101UUg9:-M1kSKdBCpzn5101UUgA:b-74u4ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/023c83e3f31201cc8e96fd7fb5f252b936f15b08/pkg/sql/colexec/execplan.go#L505)):*
> ```Go
> 		)
> 
> 		// We say that the wrapped processor is "streaming" because it is not a
> ```

nit: I'd keep this comment.
___
*[pkg/sql/colexec/execplan.go, line 898 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1kSnhWCaiHiSG40NBj:-M1kSnhWCaiHiSG40NBk:bbnrp4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/bacba6af54df86a8dfb46f75eb989eaac81b3c15/pkg/sql/colexec/execplan.go#L898)):*
> ```Go
> 			result.ColumnTypes = append(leftLogTypes, rightLogTypes...)
> 
> 			if onExpr != nil {
> ```

nit: at this point it seems worth extracting this code in a function and reusing it between hash and merge joins.
___
*[pkg/sql/colexec/mergejoiner_test.go, line 1531 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1kRU1h-IYMkWuXj1P_:-M1kRU1h-IYMkWuXj1Pa:bczgpw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/bacba6af54df86a8dfb46f75eb989eaac81b3c15/pkg/sql/colexec/mergejoiner_test.go#L1531)):*
> ```Go
> 					Inputs:               input,
> 					StreamingMemAccount:  testMemAcc,
> 					ProcessorConstructor: rowexec.NewProcessor,
> ```

I removed the unsupported join types test cases with ON expr. Is the present INNER joins cases with ON expression are unsupported? That'd be unfortunate.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45724)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colexec/execplan.go, line 1091 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1gSSPN9DzGCxC8bI3a:-M1ltOntEg2idEnQSs-Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de332a25bbbe9343e41e7a54a069b3d55fa2623/pkg/sql/colexec/execplan.go#L1091)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I see, so the actual reason is that we need to reset `.Op` and `.ColumnTypes` field to original values. And I see the value of having `reset` function right next to the struct. SGTM.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/mergejoiner_test.go, line 1531 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-M1kRU1h-IYMkWuXj1P_:-M1ltRN21CGJdo_aepT1:b6e4a7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/bacba6af54df86a8dfb46f75eb989eaac81b3c15/pkg/sql/colexec/mergejoiner_test.go#L1531)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I removed the unsupported join types test cases with ON expr. Is the present INNER joins cases with ON expression are unsupported? That'd be unfortunate.
</blockquote></details>

Done. left it in by mistake


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 6 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45724#-:-M1lviFY81d-UMFAjS50:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)



<!-- Sent from Reviewable.io -->

Commit:
colexec: wrap unsupported post process specs and ON expressions

Unsupported post process specs and ON expressions are now planned after a
columnar operator by wrapping a noop processor that evaluates the unsupported
expression before returning the row.

This commit also makes sure that two types are Equivalent before planning any
comparison or binary expression. Returning an error if not.

Release note (bug fix): mixed type comparison or binary expressions that could
have previously returned wrong results in the vectorized execution engine fall
back to using the row execution engine.
